opal/datatype/opal_datatype_position.c

```diff
@@ -61,22 +61,77 @@ position_predefined_data( opal_convertor_t* CONVERTOR,
+
+    if( cando_count > *(COUNT) )
+        cando_count = *(COUNT);
+
+    /**
+     * First check if we already did something on this element ?
+     */
+    do_now = (total_count - *(COUNT));  /* done elements */
+    if( 0 != do_now ) {
+        do_now = do_now % _elem->blocklen;  /* partial blocklen? */
+
+        if( 0 != do_now ) {
+            size_t left_in_block = _elem->blocklen - do_now;  /* left in the current blocklen */
+            do_now = (left_in_block > cando_count ) ? cando_count : left_in_block;
+            do_now_bytes = do_now * opal_datatype_basicDatatypes[_elem->common.type]->size;
+
+            OPAL_DATATYPE_SAFEGUARD_POINTER( _memory, do_now_bytes, (CONVERTOR)->pBaseBuf,
+                                            (CONVERTOR)->pDesc, (CONVERTOR)->count );
+            DO_DEBUG( opal_output( 0, "position( %p, %lu ) => space %lu [prolog]\n",
+                                   (void*)_memory, (unsigned long)do_now_bytes, (unsigned long)(*(SPACE)) ); );
+            _memory      = *(POINTER) + _elem->disp + (ptrdiff_t)do_now_bytes;
+            /* compensate if we just completed a blocklen */
+            if( do_now == left_in_block )
+                _memory += _elem->extent - (_elem->blocklen * opal_datatype_basicDatatypes[_elem->common.type]->size);
+            *(SPACE)    -= do_now_bytes;
+            *(COUNT)    -= do_now;
+            cando_count -= do_now;
+        }
+    }
+
+    /**
+     * Compute how many full blocklen we need to do and do them.
+     */
+    do_now = cando_count / _elem->blocklen;
+    if( 0 != do_now ) {
+        do_now_bytes = _elem->blocklen * opal_datatype_basicDatatypes[_elem->common.type]->size;
+        for(size_t _i = 0; _i < do_now; _i++ ) {
+            OPAL_DATATYPE_SAFEGUARD_POINTER( _memory, do_now_bytes, (CONVERTOR)->pBaseBuf,
+                                            (CONVERTOR)->pDesc, (CONVERTOR)->count );
+            DO_DEBUG( opal_output( 0, "position( %p, %lu ) => space %lu\n",
+                                   (void*)_memory, (unsigned long)do_now_bytes, (unsigned long)*(SPACE) ); );
+            _memory     += _elem->extent;
+            *(SPACE)    -= do_now_bytes;
+            *(COUNT)    -= _elem->blocklen;
+            cando_count -= _elem->blocklen;
+        }
+    }
+
+    /**
+     * As an epilog do anything left from the last blocklen.
+     */
+    do_now = cando_count;
+    if( 0 != do_now ) {
+        do_now_bytes = do_now * opal_datatype_basicDatatypes[_elem->common.type]->size;
+        OPAL_DATATYPE_SAFEGUARD_POINTER( _memory, do_now_bytes, (CONVERTOR)->pBaseBuf,
+                                        (CONVERTOR)->pDesc, (CONVERTOR)->count );
+        DO_DEBUG( opal_output( 0, "position( %p, %lu ) => space %lu [epilog]\n",
+                               (void*)_memory, (unsigned long)do_now_bytes, (unsigned long)(*(SPACE)) ); );
+        _memory   += do_now_bytes;
+        *(SPACE)  -= do_now_bytes;
+        *(COUNT)  -= do_now;
```

Comment:
Why not defining the following routine:
`static inline void process_single_block(opal_convertor_t *CONVERTOR,
                                        unsigned char *mem, ptrdiff_t mem_update,
                                        size_t *space, size_t space_update,
                                        size_t *cnt, size_t cnt_update,
                                        size_t do_now_bytes)
{
    OPAL_DATATYPE_SAFEGUARD_POINTER( mem, do_now_bytes, (CONVERTOR)->pBaseBuf,
                                    (CONVERTOR)->pDesc, (CONVERTOR)->count );
    DO_DEBUG( opal_output( 0, "position( %p, %lu ) => space %lu\n",
                           (void*)mem, (unsigned long)do_now_bytes, (unsigned long)*space ); );
    mem     += mem_update;
    *space  -= space_update;
    *cnt    -= cnt_update;
}
`

In the first check, you could call it as follows:
`process_single_block(CONVERTOR, _memory, (ptrdiff_t)do_now_bytes,
                                 SPACE, do_now_bytes, COUNT, do_now, do_now_bytes);`

In the full blocks processing, you could call it as follows:
` process_single_block(CONVERTOR, _memory, _elem->extent, SPACE, do_now_bytes, COUNT, _elem->blocklen, do_now_bytes);`

In the epilog, you could call it as follows:
`process_single_block(CONVERTOR, _memory, (ptrdiff_t)do_now_bytes, SPACE, do_now_bytes, COUNT, do_now, do_now_bytes);`